openpype/plugins/publish/integrate_new.py

```diff
@@ -110,801 +156,494 @@ class IntegrateAssetNew(pyblish.api.InstancePlugin):
-                self.integrated_file_sizes.update(self.integrate(instance))
-                self.log.debug("Integrated files {}".
-                               format(self.integrated_file_sizes))
-
-            # get 'files' info for representation and all attached resources
-            self.log.debug("Preparing files information ...")
-            representation["files"] = self.get_files_info(
-                instance,
-                self.integrated_file_sizes)
-
-            self.log.debug("__ representation: {}".format(representation))
-            destination_list.append(dst)
-            self.log.debug("__ destination_list: {}".format(destination_list))
-            instance.data['destination_list'] = destination_list
-            representations.append(representation)
-            published_representations[repre_id] = {
-                "representation": representation,
-                "anatomy_data": template_data,
-                "published_files": published_files
-            }
-            self.log.debug("__ representations: {}".format(representations))
-
-        # Remove old representations if there are any (before insertion of new)
-        if existing_repres:
-            repre_ids_to_remove = []
-            for repre in existing_repres:
-                repre_ids_to_remove.append(repre["_id"])
-            io.delete_many({"_id": {"$in": repre_ids_to_remove}})
-
-        for rep in instance.data["representations"]:
-            self.log.debug("__ rep: {}".format(rep))
-
-        io.insert_many(representations)
-        instance.data["published_representations"] = (
-            published_representations
+            # todo: reduce/simplify what is returned from this function
+            prepared = self.prepare_representation(repre,
+                                                   template_name,
+                                                   existing_repres_by_name,
+                                                   version,
+                                                   instance_stagingdir,
+                                                   instance)
+
+            for src, dst in prepared["transfers"]:
+                # todo: add support for hardlink transfers
+                file_transactions.add(src, dst)
+
+            prepared_representations.append(prepared)
+
+        if not prepared_representations:
+            # Even though we check `instance.data["representations"]` earlier
+            # this could still happen if all representations were tagged with
+            # "delete" and thus are skipped for integration
+            raise RuntimeError("No representations prepared to publish.")
+
+        # Each instance can also have pre-defined transfers not explicitly
+        # part of a representation - like texture resources used by a
+        # .ma representation. Those destination paths are pre-defined, etc.
+        # todo: should we move or simplify this logic?
```

Comment:
We've had a discussion about this internally and agreed that we should actually be adding all resources into `['fils']` list to ensure that they are correctly handled by site sync, tracked properly in the db, plus is would actually remove the need for `transfers` and `hardlinks` lists.

It would, however, require that is `representation['files']` wouldn't be only a simple list, but a list of dictionaries, to allow specifying extra information for each file. Especially considering that currently we assume everything in `files` has the same stagingDir, which is not always true`. With dictionaries we could define `file_name`, `mode`(hardlink or copy) and similar when need be.

Reply:
To be clear. 

1. Are we saying to add all files to the `instance.data["representations"]` or the actual representation entity in the database?
2. We wouldn't want database representation entity `.ma` to also have the Alembic file in `files` of representation `.abc`, correct? So we're just talking about duplicating the `files` entries for "resources" like textures in Maya Lookdev? So in essence:

```
# random example
representation .ma
    files = ["representationfile.ma", "texture1.exr", "texture2.exr"]

representation .abc
    files = ["representationfile.abc", "texture1.exr", "texture2.exr"]
```

Reply:
> We've had a discussion about this internally and agreed that we should actually be adding all resources into ['fils'] list to ensure that they are correctly handled by site sync, tracked properly in the db, plus is would actually remove the need for transfers and hardlinks lists.

@mkolar Currently I have implemented this on the database side (so the Integrator does actually integrate into `representation["files"]` in [the way I described in my previous comment](https://github.com/pypeclub/OpenPype/pull/2898#discussion_r840022014). Added with https://github.com/pypeclub/OpenPype/pull/2898/commits/89376a97e4ef85069a3afdfd5e3115b33bd27284

In essence each representation contains:
- Its own files, e.g. representation `ma` has the `.ma` and representation `abc` has the `.abc` file.
- The version's "resource" files, like textures. These are any files not explicitly assigned to a representation and considered required files for all representations. (What is currently stored in `instance.data["transfers"]` and `instance.data["hardlinks"]`)

To do is still to change how files are actually added as transfers, etc. in the extractors, etc. But currently at least the final integrator logic should now do what you expected it to do for Site Sync, etc. Yes?

Reply:
@kalisp what behaviour will this trigger with respect to site sync? If I have a file in two representations, I assume both will try to sync the file independently correct?